### PR TITLE
Add support for nested rules

### PR DESCRIPTION
**Pull request by LKay**
Adds support for validation nested objects using nested or flattened rules sets. 

Resolves issue #67.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -91,6 +91,49 @@ validation.errors.first('email'); // 'The email format is invalid.'
 validation.errors.get('email'); // returns an array of all email error messages
 ```
 
+### Nested rules
+
+Nested object could also be validated using the library. There ar two ways to declare validation rules for nested object. The first one is to declare corresponding nested object with a structure that reflects validated data structure.
```
Comment (iamdtang) [inline README.md]: Typo "ar"

File: README.md [hunk 2]
```diff
@@ -91,6 +91,49 @@ validation.errors.first('email'); // 'The email format is invalid.'
...
+var data = {
+    name: 'John',
+    bio: {
+        age: 28,
+        education: {
+            primary: 'Elementary School',
+            secondary: 'Secondary School'
+        }
+    }
+}
```
Comment (iamdtang) [inline README.md]: add semicolon

File: README.md [hunk 3]
```diff
@@ -91,6 +91,49 @@ validation.errors.first('email'); // 'The email format is invalid.'
...
+var nested = {
+    name: 'required',
+    bio: {
+        age: 'min:18',
+        education: {
+            primary: 'string',
+            secondary: 'string'
+        }
+    }
+}
```
Comment (iamdtang) [inline README.md]: please add semicolon

File: README.md [hunk 4]
```diff
@@ -91,6 +91,49 @@ validation.errors.first('email'); // 'The email format is invalid.'
...
+
+// OR
+
+var flattened = {
+    name: 'required',
+    'bio.age': 'min:18'
+    'bio.education.primary': 'string',
+    'bio.education.secondary': 'string'
+}
```
Comment (iamdtang) [inline README.md]: please add semicolon

File: package.json [hunk 5]
```diff
@@ -7,10 +7,15 @@
...
       "email": "holegary@gmail.com"
+    },
+    {
+      "name": "Karol Janyst",
+      "email": "lapkom@gmail.com"
     }
   ],
   "license": "MIT",
   "main": "./src/validator.js",
+  "typings": "validatorjs.d.ts",
```
Comment (iamdtang) [inline package.json]: could you remove typings since typescript isn't used

File: spec/flatten-object.js [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+                [{}, {}],
+                [{ foo: null }, { foo: null }],
+                [{ foo: {} }, { foo: {} }],
+                [{ foo : 1 }, { foo : 1 }],
+                [{ foo : [] }, { foo : [] }],
+                [{ foo: { bar: 1 } }, { 'foo.bar' : 1 }],
+                [{ foo: { bar: [] } }, { 'foo.bar' : [] }],
+                [{ foo: { bar: { fizz: "buzz" } } }, { 'foo.bar.fizz' : "buzz" }],
+                [{ foo: { bar: { fizz: ["buzz"] } } }, { 'foo.bar.fizz' : ["buzz"] }]
+            ]
```
Comment (iamdtang) [inline spec/flatten-object.js]: please add semicolon

File: spec/flatten-object.js [hunk 7]
```diff
@@ -0,0 +1,35 @@
...
+    var expect = require('chai').expect;
+} else {
+    var Validator = window.Validator;
+    var expect = window.chai.expect;
+}
+
+describe('Validator', function () {
+
+    context('_flattenObject', function() {
+
```
Comment (iamdtang) [inline spec/flatten-object.js]: please remove space

File: spec/flatten-object.js [hunk 8]
```diff
@@ -0,0 +1,35 @@
+if (typeof require !== 'undefined') {
+    var Validator = require('../src/validator.js');
+    var expect = require('chai').expect;
+} else {
+    var Validator = window.Validator;
+    var expect = window.chai.expect;
+}
+
+describe('Validator', function () {
+
```
Comment (iamdtang) [inline spec/flatten-object.js:26]: please remove space

File: spec/flatten-object.js (hunk 6 above)
Comment (iamdtang) [inline spec/flatten-object.js]: Looks like _flattenObject is a "private" method. If that's the case, no need to test it.

File: spec/nested-rules.js [hunk 9]
```diff
@@ -0,0 +1,90 @@
...
+
+    var dataPass = {
+        name: 'David',
+        data: {
+            weight: 70,
+            hair: {
+                color: 'black'
+            }
+        }
+    }
```
Comment (iamdtang) [inline spec/nested-rules.js:33]: please add semicolon. also, this looks like 4 spaces. could you change it all to 2 spaces?

File: validatorjs.d.ts [hunk 10]
```diff
@@ -0,0 +1,65 @@
...
+        public getRule (name: string): Rule
+        public stopOnError (passes?: Function): boolean | void
+        public passes (passes?: Function): boolean | void
+        public fails (fails?: Function): boolean | void
+
+    }
+
+    export = Validator
+
+}
```
Comment (iamdtang) [inline validatorjs.d.ts]: Let's keep this PR limited to just supporting nested rules. We can have another issue/PR for moving it over to TypeScript, which sounds like a good idea to me.

File: spec/flatten-object.js (hunk 6 above)
Comment (LKay) [inline spec/flatten-object.js]: Personally, I disagree. This method is private but any unexpected behaviour will affect class constructor or rules processing. The fact that method is private doesn't mean it doesn't have to be tested. :)

Comment (iamdtang) [inline spec/flatten-object.js]: In my opinion, testing private methods like this breaks encapsulation. We're testing it indirectly through the tests in spec/nested-rules.js. Maybe instead, you could make it its own module used by the library and test that, essentially making it public in that new module.

**Pull request merged**